src/driver/odoc_unit.ml

```diff
@@ -58,40 +60,72 @@ let of_packages ~output_dir ~linked_dir ~index_dir (pkgs : Packages.t list) :
           h pkg.libraries)
       h pkgs
   in
-  (* This one is a hashtable *)
-  let cache = Hashtbl.create 10 in
-  let pkg_args_of pkg : pkg_args =
+  let pkg_map =
+    Util.StringMap.of_list (List.map (fun pkg -> (pkg.Packages.name, pkg)) pkgs)
+  in
+
+  let lib_map =
+    Util.StringMap.of_list
+      (List.concat_map
+         (fun pkg ->
+           List.map
+             (fun lib -> (lib.Packages.lib_name, (pkg, lib)))
+             pkg.Packages.libraries)
+         pkgs)
+  in
+  let doc_dir pkg = Fpath.(pkg.Packages.pkg_dir / "doc") in
+  let lib_dir pkg libname = Fpath.(pkg.Packages.pkg_dir / "lib" / libname) in
+  let make_absolute = Fpath.( // ) output_dir in
+
+  let dash_p pkg = (pkg.Packages.name, doc_dir pkg |> make_absolute) in
+  let dash_l pkg lib =
+    (lib.Packages.lib_name, lib_dir pkg lib.lib_name |> make_absolute)
+  in
+  (* Given a pkg,  *)
+  let pkg_args_of_pkg pkg : pkg_args =
+    let own_page = dash_p pkg in
+    let own_libs = List.map (dash_l pkg) pkg.libraries in
+    { pages = [ own_page ]; libs = own_libs }
+  in
+  let pkg_args_of_config config : pkg_args =
+    let { Global_config.deps = { packages; libraries } } = config in
     let pages =
-      [
-        (pkg.Packages.name, Fpath.(output_dir // pkg.Packages.pkg_dir / "doc"));
-      ]
+      List.filter_map
+        (fun pkgname ->
+          match Util.StringMap.find_opt pkgname pkg_map with
+          | None -> None
+          | Some pkg -> Some (dash_p pkg))
+        packages
     in
     let libs =
-      List.map
-        (fun lib ->
-          ( lib.Packages.lib_name,
-            Fpath.(output_dir // pkg.Packages.pkg_dir / "lib" / lib.lib_name) ))
-        pkg.libraries
+      List.filter_map
+        (fun libname ->
+          match Util.StringMap.find_opt libname lib_map with
+          | None -> None
+          | Some (pkg, lib) -> Some (dash_l pkg lib))
+        libraries
     in
     { pages; libs }
   in
-  let pkg_args : pkg_args =
-    let pages, libs =
-      List.fold_left
-        (fun (all_pages, all_libs) pkg ->
-          let { pages; libs } = pkg_args_of pkg in
-          (pages :: all_pages, libs :: all_libs))
-        ([], []) pkgs
-    in
-    let pages = List.concat pages in
-    let libs = List.concat libs in
-    { pages; libs }
+  let pkg_args =
+    let cache = Hashtbl.create 10 in
+    fun pkg : pkg_args ->
+      match Hashtbl.find_opt cache pkg with
+      | Some res -> res
+      | None ->
+          let { pages = own_page; libs = own_libs } = pkg_args_of_pkg pkg in
+          let { pages = config_pages; libs = config_libs } =
+            pkg_args_of_config pkg.Packages.config
+          in
+          { pages = own_page @ config_pages; libs = own_libs @ config_libs }
```

Comment:
These will be used for both pages and modules ? I think that's fine. My initial idea was that the new dependencies will be specified on the `(documentation)` stanza and it might be weird that they also affect modules. It seems that this will be defined in `dune-project` instead, which is fine.